elastic/kibana
# [App Search] Fixed 2 relevance tuning bugs

## Summary

Bug 1: Selecting a "Functional" boost for a "Date" field will result in a "Functional boosts are not supported for Date fields..." error message.

The fix for that is here: https://github.com/elastic/kibana/commit/ea0a3afdc2962a9fe8f232d8acd1bb1538773fd8

To Test:
- Create an engine that has a "Date" field in its schema
- Navigate to Your engine > Relevance Tuning > your date field > "Add Boost" > you should not see "Functional" listed as an option, only "Value" and "Proximity"

Bug 2: Selecting a "Functional" or "Proximity" boost and pressing save, without selecting an option, was causing an error.

The root cause was that when creating new Functional and Proximity boosts in state, we were not initializing all required fields of the boost. So when we attempt to post it to the server, the server was rejecting it as invalid

As it turns out, the reason we missed this on the initial implementation, is because that initialization was actually happening in the view, rather than Kea in the old repository. I.e., when the view was initialized, it was actually mutating the newly created boosts in state directly, rather than using a kea action. This is a pretty clear anti-pattern. My fix for this is to initialize the missing values as part of the "addBoost" action in Kea: https://github.com/elastic/kibana/commit/05ccd1868eb87ffc3d7510b0aeb63c20c5634752. https://github.com/elastic/kibana/pull/94312/files#diff-8f932314a95e85fb9e58bdfb28af437b2d3be46a00caa15052d09619f96bf26eR373

Note that as part of this, I refactored the Boost types a bit to make the schema more clear. There is now a parent Boost type as well as child FunctionalBoost, ValueBoost, and ProximityBoost types, to make it clear which Boost fields are used for which boost types.

To test:
- Create an engine that has a "Number" field in its schema
- Navigate to Your engine > Relevance Tuning > your number field > "Add Boost" > "Functional"
- Observe that the Functional boost is created in the UI with the dropdowns pre-populated.
- Use the "Save" button to save the new boost.
- You should not see an error.
- You should see that the Functional boost is still on the page and was persisted.
- Repeat the previous steps for a Proximity boost (note that you will need to enter a "center" value in order to save this one properly). The center value is not pre-populated.


### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
